confidential-datahub/docs/SEALED_SECRET.md

```diff
@@ -0,0 +1,153 @@
+# Kubernetes Sealed Secret
+
+## Introduction
+
+CoCo’s [threat model](https://github.com/confidential-containers/confidential-containers/blob/main/trust_model_personas.md)
+excludes the Kubernetes Control Plane and Host components from the
+Trusted Compute Base (TCB). With this stricter threat model, CoCo
+runtimes cannot take advantage of the existing infrastructure. to
+protect [Kubernetes secrets](https://kubernetes.io/docs/concepts/configuration/secret/),
+this document introduces Sealed Secrets that supports in-pod decryption.
+
+A Sealed Secret is a set of metadata to get the plaintext of the secret.
+A Sealed Secret will be unsealed inside the TEE pod transparently to users,
+and the unsealing occurs only if the remote attestation process passes,
+which means the TEE environment is as expected. Also, Sealed Secret can
+leverage commercial KMS/Secret Manager(Vault) productions in the unsealing
+process.
```

Comment:
Not yet. However, after merging this we will make an PR that defines the trait of a KMS and a Vault. Semantically there will be two main apis:
```
async fn decrypt(&mut self, ciphertext: &[u8], keyid: &str) -> Result<Vec<u8>>;
async fn get_secret(&mut self, name: &str) -> Result<Vec<u8>>;
```

Thus, the code of Sealed Secret will just call the trait function to get the secret directly or unseal the wrapped secret, in this way we do not need to care much about the concrete vault/kms underneath.

I glanced at [api of hashicorp vault](https://docs.rs/hashicorp_vault/2.1.1/hashicorp_vault/client/struct.VaultClient.html#method.get_secret) and found it matches the definition of `get_secret()` so it can support `Vault` type sealed secret. I'm not sure whether `decrypt` like api is supported by hashicorp vault. If so, `KMS`(or we say `Envelope`) type Sealed Secret can be supported.